sqsconsumer.go

```diff
@@ -178,50 +182,64 @@ func (m *SmartSQSConsumer) StartConsuming(ctx context.Context) error {
 
 // worker function represents a single "message worker." worker will infinitely process messages until
 // messages is closed. worker is responsible for handling deletion of messages, or handling
-// messages that have retryable error.
+// messages that have retryable error. On the event of a retryable error, worker will change
+// the visibilitytimeout of a message to be the configured retryableErr.VisibilityTimeout
 func (m *SmartSQSConsumer) worker(ctx context.Context, messages <-chan *sqs.Message) {
 	for message := range messages {
-		err := m.GetSQSMessageConsumer().ConsumeMessage(ctx, message)
-		if err != nil {
-			switch err.(type) {
-			case RetryableConsumerError:
-				retryableErr := err.(RetryableConsumerError)
-				m.ackMessage(ctx, func() error {
-					var _, e = m.Queue.ChangeMessageVisibility(&sqs.ChangeMessageVisibilityInput{
-						QueueUrl:          aws.String(m.QueueURL),
-						ReceiptHandle:     message.ReceiptHandle,
-						VisibilityTimeout: &retryableErr.VisibilityTimeout,
+		consumerErr := m.GetSQSMessageConsumer().ConsumeMessage(ctx, message)
+		if consumerErr != nil {
+			if consumerErr.IsRetryable() {
+				// check to see if we've reached the maximum number of retries allowed
+				receiveCount := getApproximateReceiveCount(message)
```

Comment:
Suggestion: relying on `ApproximateReceiveCount` for retrying may not be reliable. SQS is atleast once delivery which will skew the count. Also, if any other system or console reads SQS message the count will get incremented.